`BaseRestartWorkChain`: allow to override priority in `handler_overrides`

sphuber opened the pull request:
Fixes #5545 

The `handler_overrides` could so far be used to override the `enabled`
keyword of the corresponding handler. This would allow to disable or
enable a handler on a per process instance basis.

A user may want to do the same for the `priority`. To make this possible
the type of the `handler_overrides` is changed where the values should
now be dictionaries where the keys `enabled` and `priority` are
supported. These can be used to override the original values declared
in the source code of the work chain.

To provide backwards-compatibility, the old syntax is still supported
and automatically converted, with a deprecation warning being displayed.

sphuber commented:
Still want to add docs to this as the `handler_overrides` are not documented at all currently I believe.

espenfl commented:
@sphuber Thanks a lot. Happy to add docs. Where should it go?

espenfl commented on aiida/engine/processes/workchains/restart.py:
```diff
@@ -31,32 +32,44 @@ def validate_handler_overrides(
-    instance method of the `process_class` that has been decorated with the `process_handler` decorator. The values
-    should be boolean.
+    The ``handler_overrides`` should be a dictionary where keys are strings that are the name of a process handler, i.e.
+    an instance method of the `process_class` that has been decorated with the `process_handler` decorator. The values
```
Should `process_handler` still be in single quotes?

sphuber replied:
No, missed that one. Fixed now.

sphuber commented:
Thanks for the review @espenfl . I actually noticed there was a subtle bug that wasn't being tested for. The initial implementation would actually affect the handlers for multiple executions and not just the single instance. This is now fixed and explicitly tested. I have also added the docs.
